libass/ass_render.c

```diff
@@ -662,7 +662,8 @@ static void blend_vector_clip(ASS_Renderer *render_priv, ASS_Image *head)
     ol_key.u.drawing.text = render_priv->state.clip_drawing_text;
 
     double m[3][3] = {{0}};
-    double w = render_priv->font_scale / (1 << (render_priv->state.clip_drawing_scale - 1));
+    int32_t scale_base = lshiftwrapi(1, render_priv->state.clip_drawing_scale - 1);
+    double w = render_priv->font_scale / (scale_base ? scale_base : 1);
```

Comment:
Here and below, it may make more sense to set `w = 0` or something if `!scale_base`, because I imagine VSFilter wouldn’t display anything for such an item, as it would end up with every single coordinate inside the drawing equal to `INT32_MIN`, reducing the drawing to an infinitesimal, off-screen point.

Reply:
I haven't checked the following calculations too closely, but using `0` seems sensible so it does now.

Reply:
Do we even need to emulate never used VSFilter bugs? It's ugly cumbersome code without any tangible benefit. Also, it's compatibility not with some existing VSFilter source (as it's UB that should be fixed there too), but with compiled binary with some particular behavior for this UB. I think that
```c
double w = ldexp(priv->font_scale, 1 - info->drawing_scale);
```
is much cleaner and more intuitively correct in this case.

Reply:
How did you conclude that this particular quirk is “never used”? I'm inclined to default to sticking closely to VSFilter, since there are many weird VSFilter quirks relied upon in real-world scripts and emulating this is sufficiently easy *(and I expect it to probably not negatively effect sane and otherwise VSFilter-compatible scripts)*.

Reply:
This quirk doesn't introduce any new behavior that can be useful, hard to trigger accidentally, and no authoring/automation tool output it (actually anything except `\p1` is outdated by floating-point coordinates extension). So there's basically no way it can creep into real subs outside of fuzzing. There's already a ton of widely used VSFilter bugs, so I think it's better to focus on that than trying to emulate every possible UB from its source code.

Reply:
Sufficiently high enough drawscale values can easily occur if an additional key was hit by accident while typing. How common this is or not and the remains unnoticed with VSFilter- and “current-libass on x86”-behaviour, I do not know.  After some rounds of simplification, the emulation code is now just one line. And since the effects of drawscale values higher than 31 can also be achieved by decimal coordinates as you pointed out, sub authors don't gain anything useful by us not emulating VSFilter here, but everyone will end up with a new compatibility concern/issue. Thus I believe explicitly emulating VSF on leftshifts is the better choice.